sam/header.go

```diff
@@ -126,12 +134,85 @@ func NewHeader(text []byte, r []*Reference) (*Header, error) {
 	return bh, nil
 }
 
+// Get returns the string representation of the value associated with the
+// given header line tag. If the tag is not present the empty string is returned.
+func (bh *Header) Get(t Tag) string {
+	switch t {
+	case versionTag:
+		return bh.Version
+	case sortOrderTag:
+		return bh.SortOrder.String()
+	case groupOrderTag:
+		return bh.GroupOrder.String()
+	}
+	for _, tp := range bh.otherTags {
+		if t == tp.tag {
+			return tp.value
+		}
+	}
+	return ""
+}
+
+// Set sets the value associated with the given header line tag to the specified
+// value. If value is the empty string and the tag may be absent, it is deleted
+// or set to a meaningful default (SO:UnknownOrder and GO:GroupUnspecified),
+// otherwise an error is returned.
+func (bh *Header) Set(t Tag, value string) error {
+	switch t {
+	case versionTag:
+		if value == "" {
+			return errBadHeader
+		}
+		bh.Version = value
+	case sortOrderTag:
+		if value == "" {
+			bh.SortOrder = UnknownOrder
+			return nil
+		}
+		sortOrder, ok := sortOrderMap[value]
+		if !ok {
+			return errBadHeader
+		}
+		bh.SortOrder = sortOrder
+	case groupOrderTag:
+		if value == "" {
+			bh.GroupOrder = GroupUnspecified
+			return nil
+		}
+		groupOrder, ok := groupOrderMap[value]
+		if !ok {
+			return errBadHeader
+		}
+		bh.GroupOrder = groupOrder
+	default:
+		if value == "" {
+			for i, tp := range bh.otherTags {
+				if t == tp.tag {
+					copy(bh.otherTags[i:], bh.otherTags[i+1:])
+					bh.otherTags = bh.otherTags[:len(bh.otherTags)-1]
+					return nil
+				}
+			}
+		} else {
+			for i, tp := range bh.otherTags {
+				if t == tp.tag {
+					bh.otherTags[i].value = value
+					return nil
+				}
+			}
+			bh.otherTags = append(bh.otherTags, tagPair{tag: t, value: value})
+		}
+	}
+	return nil
+}
+
 // Clone returns a deep copy of the receiver.
 func (bh *Header) Clone() *Header {
 	c := &Header{
 		Version:    bh.Version,
 		SortOrder:  bh.SortOrder,
 		GroupOrder: bh.GroupOrder,
+		otherTags:  append([]tagPair(nil), bh.otherTags...),
 		Comments:   append([]string(nil), bh.Comments...),
```

Comment:
I haven't seen this idiom, it just allows you to create (and append to) an empty slice in-place?